### PR TITLE
Improve the SineReLU activation function

### PR DESCRIPTION
**Pull request by wilderrodrigues**
Hi @fchollet and @ahundt,

I put some more time in the SineReLU and improved it, from a performance and running time perspective. The function is now simpler, faster and the results are better.

* Increase speed by reducing the computation: instead of using sigmoid and e, just use the difference of sin(Z) and cos(Z).
* Make the function differentiable

I wrote a Medium post about it, which will serve as inspiration for a paper to, hopefully, get out next year (which will be way more technical than the post).

Next to the documentation written in the code, I also took the time run it several times and get a more interesting comparison:

image

Link for the Medium story: https://medium.com/@wilder.rodrigues/sinerelu-an-alternative-to-the-relu-activation-function-e46a6199997d

Thanks in advance.

### CODE REVIEW

**Comment by wilderrodrigues** [comment]
Having a look at the failed check.

**Comment by wilderrodrigues** [comment]
It is not related, but I will try to get that test fixed:

```
ImportError while importing test module '/home/travis/build/keras-team/keras-contrib/tests/keras_contrib/utils/save_load_utils_test.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/keras_contrib/utils/save_load_utils_test.py:9: in <module>
    from keras_contrib.utils.save_load_utils import save_all_weights, load_all_weights
keras_contrib/utils/save_load_utils.py:7: in <module>
    from keras.legacy import models as legacy_models
E   ImportError: cannot import name models
```

**Comment by wilderrodrigues** [comment]
Okay, just saw on the other PR that @ahundt and @titu1994 are busy with it. Will see how I can help.

**Comment by ahundt** [comment]
cool, looks like a decent change, though do those MNIST numbers mean a lot?
Have you tried Fashion-MNIST? It is a more difficult example already in keras with a data format binary equivalent to MNIST, so that should be a one liner change.

Travis should be fixed soon, then you'll need to merge to get it to pass.

**Comment by titu1994** [comment]
@wilderrodrigues Just had a look at the changes. LGTM.

I was wondering about something. Have you ever tried to let it learn the scaling factor `epsilon` on its own? As in treat it as a trainable weight, with a small enough initialization (say 1e-3 or 2.5e-3) and let it be optimized by gradient descent directly so as to avoid a hand tuned epsilon?

Of course, a multiplicative parameter is harder to optimize so some regularization has to be placed to avoid poor training in the initial epochs, or even clip to certain ranges to avoid erratic gradient behaviour.

Simply put, the choice of `epsilon` seems to be significant to the working of the activation (due to it being multiplied by the magnitude of the activation), and while hyperopt is fine to select good values on smaller datasets, having to select an optimal value for an activation function reduces its effectiveness significantly (where one can get nearly same result by just choosing to use ReLU and avoid a significant search space altogether).

**Comment by wilderrodrigues** [comment]
Hi @ahundt and @titu1994,

Thanks for taking the time to look at the changes and also make a couple of suggestions. Some notes below:

1. I haven't tried with Fashion-MNIST yet. But will give it a go for sure. I'm busy writing a paper about the function in an attempt to spread it around quicker. So, Fashion-MNIST will be one of the data sets for sure, along with ImageNet and CIFAR-10/100.

2. That's a cool idea, @titu1994. It did not come to my mind before and will certainly be part of the new pull request if I get it to work nicely. 

Quick question from my side: can this PR be merged or would the changes you suggested be required before the merge? I can work on them in a separate PR, no problem.

Thanks again!

Cheers,
Wilder

**Comment by titu1994** [comment]
@wilderrodrigues I think this PR is sufficient for now. Fashion MNIST results can be added to another PR at a later date, and the suggestion I made requuress a whole slew of new experiments which are unrelated to this PR at the moment. 
@ahundt will have final say on whether this get merged as is or with the fashion MNIST

**Comment by ahundt** [comment]
Removing sigmoid will improve the gradients anyway, so I will merge this improvement now. 

How fast/slow is this activation? Don't worry if you don't have the time, but if you run another experiment, could you also get the runtime compared to relu in addition to the accuracy? Simply running `time python mytest.py` in front of the python code could do the trick or using the `timeit` library.

**Comment by wilderrodrigues** [comment]
Hi @ahundt,

I have some time benchmark from TensorBoard, but I can run another experiment and use `timeit`. I will take some time tomorrow to do it.

Cheers,
Wilder

**Comment by wilderrodrigues** [comment]
I'm running a deep net (4 layers) using Fashion MNIST now. Trying it out on my macbook, but will get a cluster with it running a more robust model (CNN) and will create a PR with results (including runtime) by next week.

**Comment by titu1994** [comment]
@wilderrodrigues I suggest using [Colaboratory by Google](https://colab.research.google.com) with a GPU attached for quick experimentation and then export the notebook. You can pip install your branch of Keras contrib in order to access the activation function directly.

**Comment by wilderrodrigues** [comment]
Hi @ahundt and @titu1994,

I got some results running the Fashion MNIST with Google Colab and IBM Data Science Experience Platform (both with free GPUs). I also benchmarked against the LeakyReLU, so we have a better idea about the SineReLU's performance. However, I haven't tested with CNNs on those platforms, but will also do so.

* Network Architecture and Hyper Parameters: 3 Hidden layers (128, 512, 1024); Dropout (.2, .3, .5); Xavier Glorot initialiser; Adam optimiser; 50 epochs.

* Software:
  * `# !pip install --upgrade git+https://www.github.com/ekholabs/keras-contrib.git@improve-sinerelu`
  * `# !pip install --upgrade keras==2.2.0`
  * `# !pip install --upgrade tensorflow==1.8.0`

* Google Colab:
```
#SineReLU

#Final loss: 0.3392, final accuracy: 0.8947 -> SineReLU (0.006); STD loss => 0.08377053628635789
#Final loss: 0.3604, final accuracy: 0.8941 -> SineReLU (0.006); STD loss => 0.08358033621210441
#Final loss: 0.3517, final accuracy: 0.8911 -> SineReLU (0.006); STD loss => 0.08390637325259301
#Final loss: 0.3515, final accuracy: 0.8908 -> SineReLU (0.006); STD loss => 0.08358472175590163
#Final loss: 0.3548, final accuracy: 0.8903 -> SineReLU (0.006); STD loss => 0.08382269174213763
#Final loss: 0.3561, final accuracy: 0.8900 -> SineReLU (0.006); STD loss => 0.08384760879705476

#LeakyReLU

#Final loss: 0.3529, final accuracy: 0.8942 -> LeakyReLU (0.01); STD loss => 0.08346229346258482
#Final loss: 0.3474, final accuracy: 0.8922 -> LeakyReLU (0.01); STD loss => 0.08294998525464506
#Final loss: 0.3368, final accuracy: 0.8908 -> LeakyReLU (0.01); STD loss => 0.08321019835026658
#Final loss: 0.3618, final accuracy: 0.8889 -> LeakyReLU (0.01); STD loss => 0.08315353110347731
#Final loss: 0.3770, final accuracy: 0.8872 -> LeakyReLU (0.01); STD loss => 0.08327670884924857
#Final loss: 0.3561, final accuracy: 0.8859 -> LeakyReLU (0.01); STD loss => 0.0826444040529247

#ReLU

#Final loss: 0.3617, final accuracy: 0.8911; STD loss => 0.08343571727475199
#Final loss: 0.3598, final accuracy: 0.8907; STD loss => 0.08402013962686715
#Final loss: 0.3421, final accuracy: 0.8905; STD loss => 0.08357489313026156
#Final loss: 0.3382, final accuracy: 0.8902; STD loss => 0.08356168059998799
#Final loss: 0.3563, final accuracy: 0.8851; STD loss => 0.08340573980083979
#Final loss: 0.3534, final accuracy: 0.8828; STD loss => 0.08353081964537325
```

* IBM Data Science Experience Platform
```
#SineReLU

#Final loss: 0.3514, final accuracy: 0.8946; SineReLU (0.006); STD loss => 0.0845423117762957
#Final loss: 0.3466, final accuracy: 0.8940; SineReLU (0.006); STD loss => 0.08401259096681588
#Final loss: 0.3675, final accuracy: 0.8920; SineReLU (0.006); STD loss => 0.08434074186936903
#Final loss: 0.3432, final accuracy: 0.8916; SineReLU (0.006); STD loss => 0.08423367069518602
#Final loss: 0.3521, final accuracy: 0.8892; SineReLU (0.006); STD loss => 0.08462476684233458

#LeakyReLU

#Final loss: 0.3497, final accuracy: 0.8944; LeakyReLU (0.01); STD loss => 0.08339076990387603
#Final loss: 0.3531, final accuracy: 0.8915; LeakyReLU (0.01); STD loss => 0.08295369680579048
#Final loss: 0.3542, final accuracy: 0.8907; LeakyReLU (0.01); STD loss => 0.08355390322983067
#Final loss: 0.3615, final accuracy: 0.8852; LeakyReLU (0.01); STD loss => 0.08331255171834445
#Final loss: 0.3793, final accuracy: 0.8839; LeakyReLU (0.01); STD loss => 0.08320457550420218

#ReLU

#Final loss: 0.3459, final accuracy: 0.8937; STD loss => 0.08364857031790071
#Final loss: 0.3462, final accuracy: 0.8914; STD loss => 0.08330743281652375
#Final loss: 0.3486, final accuracy: 0.8903; STD loss => 0.08312365975992159
#Final loss: 0.3540, final accuracy: 0.8894; STD loss => 0.08356918250251755
#Final loss: 0.3626, final accuracy: 0.8889; STD loss => 0.08400160885157779
```

Question: should I add all this to the Python docstrings? I think it would be a bit too much, right?

The CNN architecture takes a bit more of time, but I will run it a few times with the three activation functions on the IBM platform (faster than Colab).

I hope this is sufficient for the merge. I will work on the Adaptive Learning Epsilon in a separate branch. 

Thanks again for your support!

**Comment by titu1994** [comment]
With pip install Tensorflow you are installing the CPU version of Tensorflow, so that could account for your CNN model taking so much time. 

In any case, stating the mean + standard deviation of the above on six lines is more than sufficient to merge here. 

@ahundt Any comments ?

**Comment by wilderrodrigues** [comment]
Yep, you are right! I have to do a `pip install tensorflow-gpu`. Done that several times for `nvidia-docker` but now just forgot. :(

Will fix and restart.

**Comment by ahundt** [comment]
I think @titu1994 made perfect comments, looks good. As he said you just need a one line comment for each case & include with a mean + std dev.

**Comment by wilderrodrigues** [comment]
Thanks, guys!

I will add a comment in the docstring with the mean of the accuracy and the mean of standard deviation of the loos.

I got the GPU to run fine on Googla Colaboratory and the CNN is taking only 25 seconds per epoch. So, will run it on the three functions and add the outcome to the docstring as well.

Cheers,
Wilder

**Comment by wilderrodrigues** [comment]
Hi @ahundt and @titu1994,

I added the benchmark results after running a fully connected and convolutional neural networks agains both Fashion MNIST and MNIST. I did 6 consecutive runs with each architecture, cleaning the state before each run, using the SineReLU, ReLU and LeakyReLU.

The metrics consist of:

* Mean of the loss
* Mean of the accuracy
* Mean of the standard deviation of the loss

Please, do not hesitate to let me know if you need more details.

Cheers,
Wilder

**Comment by titu1994** [comment]
Thanks for the PR !
